Added: Full TCP throughput

monfron opened the pull request:
- TCP throughput now works: i.e. the control plane can handle the commands received from client, the measurement module then process
- handle several concurrent measurmeents
- many more stuff.

- what does NOT work:
- some fields are still hardcoded
- QUIC and UDP are still dummys (that means the general functionality works but NO NETCODE)

monfron committed:
Refactoring: byteRecvd class attribute
- Note: when we're finisihed we have to change that for the
UDP module aswell
- Note2: As a result the server does not display the
currently received bytes
Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: "Data" field in DataObj

-this data field holds an array of DataElements
- so we get the desired layout of the Mapago STD
Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Implemented: ctrl plane can handle msmstInfoRep, ch handling, etc.
- That means: We can handle Msmt Info Rep
- We can construct a MsmtInfoReply

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Implemented: Client sends MsmtInfoRequest

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Implement: Per-Stream-Byte-Counter, Mutex for map
- Note: timestamps are still dummys

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: srv-tcp-worker selects listen port, TODOs, gofmt

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: Conv int-slice (used ports) to string (json)

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Improved: DataObj struct now includes usedPorts

- Note: honestly i dont know how we can transmit a empty interface including
a) a STRING MAP and b) DATA RESULT struct.

=> we can only perform one type assertion....

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Implement: full server-decides-listen-port-functionality

NOTE: UDP and QUIC are still dummys regarding the MEASUREMENT PLANE,
the CONTROL PLANE can handle everything

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

hgn commented on control-plane/control-plane.go:
```diff
@@ -210,6 +233,28 @@ func constructMsmtStopReply(reqDataObj *shared.DataObj, msmtRep shared.ChMsmt2Ct
+func constructMsmtInfoReply(reqDataObj *shared.DataObj, msmtRep shared.ChMsmt2Ctrl) *shared.DataObj {
+	fmt.Println("\n------------- Constructing MSMT_INFO_REPLY -------------")
+
+	repDataObj := new(shared.DataObj)
```
alles ist ein "object" das kannste fast überall entfernen.

monfron replied:
mach ich wenn ich zeit hab...ich mach jetzt erst mal weiter

hgn commented on control-plane/control-plane.go:
```diff
@@ -210,6 +233,28 @@ func constructMsmtStopReply(reqDataObj *shared.DataObj, msmtRep shared.ChMsmt2Ct
+	repDataObj.Id = ID
+	repDataObj.Seq_rp = reqDataObj.Seq
+
+	msmtData, ok := msmtRep.Data.([]shared.DataResultObj)
```
mhh, das kann doch passieren, oder? Warum gleich das Programm abbrechen?

monfron replied:
naja ich will mich doch hier vergewissern dass das empty interface ein slice von den results ist

hgn commented on control-plane/ctrl/shared/ctrl-msg.go:
```diff
@@ -61,23 +62,24 @@ type MsmtStorageEntry struct {
+	Padding              string            `json:",omitempty"`
+	Modules              string            `json:",omitempty"`
+	Measurement          MeasurementObj    `json:",omitempty"`
+	Data                 DataCollectionObj `json:",omitempty"`
```
müssen alle Felder immer gesetzt sein? Wo ist die Trennung zu finden ob etwas optional ist als mandatory? Im Code wahrscheinlich, oder?

monfron replied:
ne müssen ned gesetzt sein. wenn das feld nicht belegt wird dann wirds auch nicht ins json gepackt

hgn commented on measurement-plane/tcp-throughput/server.go:
```diff
@@ -39,15 +34,15 @@ type TcpMsmtObj struct {
 }
 
-func NewTcpMsmtObj(msmtCh <-chan shared.ChMgmt2Msmt, ctrlCh chan<- shared.ChMsmt2Ctrl, msmtStartReq *shared.DataObj, msmtId string) *TcpMsmtObj {
+func NewTcpMsmtObj(msmtCh <-chan shared.ChMgmt2Msmt, ctrlCh chan<- shared.ChMsmt2Ctrl, msmtStartReq *shared.DataObj, msmtId string, startPort int) *TcpMsmtObj {
```
go format?

monfron replied:
hab ich gemacht

hgn commented on measurement-plane/tcp-throughput/server.go:
```diff
@@ -113,32 +98,33 @@ func NewTcpMsmtObj(msmtCh <-chan shared.ChMgmt2Msmt, ctrlCh chan<- shared.ChMsmt
+			tcpMsmt.usedPorts = append(tcpMsmt.usedPorts, port)
+			goHeartbeatCh <- true
+			break
+		} else {
```
else kannst du dir hier sparen, breaked ja im error case bereits

monfron replied:
stimmt

hgn commented on measurement-plane/tcp-throughput/server.go:
```diff
@@ -218,32 +198,34 @@ func (tcpMsmt *TcpMsmtObj) CloseConn() {
+
+		// these are dummy values
+		dataElement.Timestamp_first = shared.ConvCurrDateToStr()
+		dataElement.Timestamp_last = shared.ConvCurrDateToStr()
```
dataElement.Timestamp_last = dataElement.Timestamp_first;

hgn commented:
Wo ist der port code?

hgn commented:
Was muss ich machen um das selber mal bei mir zu testen? einfach die zwei befehle die ich nach dem kompilieren benötige um client und server zu starten. Dann kann ich das mal checken!

monfron committed:
Fixed: Small adjustment

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: Correct timestamps for msmt-data

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: udp-throughput/server.go (class layout, complete functionality)
- the udp-throughput module is now shifted towards the tcp class layout
- NOTE: the management plane cant handle the class-based functionality ATM
- the next commit will implement that

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Removed: Old udp-msmt handling, Added: Server fully supports udp-msmt class handling

Signed-off-by: monfron <simon.fronhoefer@gmail.com>
